### PR TITLE
Feature/ont 44

### PR DESCRIPTION
**Pull request by ctucker3**
Implements POST traits endpoint. This card was updated to just be the backend work. 

## Related work

- POST programs needed modification to create the program ontology and a program observation level when a new program was created. 
- BrAPI test server was added to our project
- The way we spin up test contains was changed to allow for a temporary BrAPI test server to be spun up for testing 
- Trait upload was changed slightly to use the shared trait validation
- Trait validation was updated to return multiple error responses
- Our new brapi test server image is on docker hub, https://hub.docker.com/repository/docker/breedinginsight/brapi-java-server
- Our new brapi test server repository is in github, https://github.com/Breeding-Insight/brapi-Java-TestServer

## Questions for reviewer

- I don't like the way that the `column` field in the multi-error response is populated. Right now it is a manually specified string that uses dot syntax to indicate hierarchy (ex. `traitName` and `method.methodClass`). I wonder if there is a better way to do this so if our field name changes we either don't need to change these strings, or they will error out so we need to change them. 

## Work we need to do

- I created a program observation level of "Plant" for any new program created. This will need to be replaced when we have the ability to create program observation levels. This might hold up the science team using the trait upload fully because you cannot upload a trait without a proper program observation level.

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -25,7 +25,14 @@ API_INTERNAL_PORT=8081
 API_INTERNAL_TEST_PORT=8082
 
 # Brapi Server Variables
-BRAPI_CORE_URL=https://test-server.brapi.org/
-BRAPI_PHENO_URL=https://test-server.brapi.org/
-BRAPI_GENO_URL=https://test-server.brapi.org/
-BRAPI_REFERENCE_SOURCE=breeding-insight.org
\ No newline at end of file
+BRAPI_CORE_URL=http://localhost:8083/
```
Comment (nickpalladino) [inline .env:28]: should this be pre-defined here as localhost:8083 if it depends on the brapi configuration below? Should it use the BRAPI_SERVER_PORT variable instead?

File: docs/openapi.yaml [hunk 2]
```diff
@@ -2214,13 +2251,13 @@ paths:
                 type: string
               example: ERROR - 2018-10-08T20:15:11Z - The requested object DbId is
                 not found
-  /topography-options:
+  /topographies:
```
Comment (nickpalladino) [inline docs/openapi.yaml:2254]: Why did these location information endpoints change, maybe a merging or out of date issue? They don't match the controller specified endpoints?

File: docs/openapi.yaml [hunk 3]
```diff
@@ -1939,9 +1979,6 @@ paths:
                 file:
                   type: string
                   format: binary
-            encoding:
```
Comment (nickpalladino) [inline docs/openapi.yaml:1942]: Why was this removed? Probably related to the above changes

File: .env (hunk 1 above)
Comment (ctucker3) [reply .env:28]: Yeah I like that idea!

File: src/main/java/org/breedinginsight/services/TraitService.java [hunk 4]
```diff
@@ -59,4 +83,107 @@
 
        return traitDAO.getTraitFull(programId, traitId);
     }
+
+    public List<Trait> createTraits(UUID programId, List<Trait> traits, AuthenticatedUser actingUser)
+            throws DoesNotExistException, ValidatorException {
+
+        Optional<Program> optionalProgram = programService.getById(programId);
+        if (!optionalProgram.isPresent()) {
```
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/services/TraitService.java:91]: you could get fancy on these optional checks if you want: 
```Program program = optionalProgram.orElseThrow(() -> new  DoesNotExistException("Program does not exist"));```

File: docs/openapi.yaml (hunk 2 above)
Comment (ctucker3) [reply docs/openapi.yaml:2254]: I didn't intentionally changes these, so probably one of the issues you mentioned, or a copy and paste issue! I'll look into it.

**Comment by nickpalladino** [comment]
[ONT-44 - Finalize trait import](https://trello.com/c/B6sUWTB3/246-ont-44-finalize-trait-import)

File: src/main/java/org/breedinginsight/daos/TraitDAO.java [hunk 5]
```diff
@@ -181,6 +202,130 @@ public TraitDAO(Configuration config, DSLContext dsl) {
         return Optional.of(parseTraitRecord(record, createdByUser, updatedByUser));
     }
 
+    public List<Trait> createTraitsBrAPI(List<Trait> traits, User actingUser, Program program){
+
+        //TODO: Pass ontology reference
```
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/daos/TraitDAO.java:207]: Do we have a card for creating the ontology and using the reference?

File: src/main/java/org/breedinginsight/daos/TraitDAO.java [hunk 6]
```diff
@@ -181,6 +202,130 @@ public TraitDAO(Configuration config, DSLContext dsl) {
...
+                    //.contextOfUse()
+                    //.documentationURL()
+                    //.growthStage()
+
+            brApiVariables.add(brApiVariable);
+        }
+
+
+        // POST variables to each brapi service
+        // TODO: If there is a failure after the first brapi service, roll back all before the failure.
```
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/daos/TraitDAO.java:292]: Do we have a card for api roll back and more specifically applying it here when we have it?

Comment (ctucker3) [reply src/main/java/org/breedinginsight/daos/TraitDAO.java:292]: https://trello.com/c/sZmjsdYW/37-acid-transactions-for-brapi-calls

File: src/main/java/org/breedinginsight/daos/TraitDAO.java (hunk 5 above)
Comment (ctucker3) [reply src/main/java/org/breedinginsight/daos/TraitDAO.java:207]: https://trello.com/c/4V5bXNBb/166-ont-59-create-brapi-ontology-on-program-create

**Comment by ctucker3** [comment]
Note on my question about a better way to do column names:

I'm working on an enum for the trait column names that are used in the validators to facilitate mapping back to the trait file column names in my work for ONT-68.

File: src/main/java/org/breedinginsight/services/TraitService.java [hunk 7]
```diff
@@ -59,4 +83,107 @@
...
+
+        // Validate the traits
+        ValidationErrors requiredFieldErrors = TraitValidator.checkRequiredTraitFields(traits);
+        ValidationErrors dataConsistencyErrors = TraitValidator.checkTraitDataConsistency(traits);
+        ValidationErrors duplicateTraits = traitValidator.checkDuplicateTraitsExisting(traits);
+        ValidationErrors duplicateTraitsInFile = TraitValidator.checkDuplicateTraitsInFile(traits);
+        validationErrors.mergeAll(requiredFieldErrors, dataConsistencyErrors, duplicateTraits, duplicateTraitsInFile);
+
+        if (validationErrors.hasErrors()){
+            throw new ValidatorException(validationErrors);
```
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/services/TraitService.java:139]: If I pass in an empty json array or even empty json object in the body of the request, I get a 422 with the following output: 
```
{
  "rowErrors": [
    {
      "rowIndex": 0,
      "errors": [
        {
          "column": "method",
          "errorMessage": "Missing method",
          "httpStatus": "UNPROCESSABLE_ENTITY",
          "httpStatusCode": 422
        },
        {
          "column": "scale",
          "errorMessage": "Missing scale",
          "httpStatus": "UNPROCESSABLE_ENTITY",
          "httpStatusCode": 422
        },
        {
          "column": "traitName",
          "errorMessage": "Missing trait name",
          "httpStatus": "UNPROCESSABLE_ENTITY",
          "httpStatusCode": 422
        },
        {
          "column": "description",
          "errorMessage": "Missing trait description",
          "httpStatus": "UNPROCESSABLE_ENTITY",
          "httpStatusCode": 422
        },
        {
          "column": "programObservationLevel.name",
          "errorMessage": "Missing trait level",
          "httpStatus": "UNPROCESSABLE_ENTITY",
          "httpStatusCode": 422
        }
      ]
    }
  ]
}
```

I think it should be a 400 like you were talking about.

File: src/main/java/org/breedinginsight/services/TraitService.java (hunk 4 above)
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/TraitService.java:91]: You keep your voodoo magic out of my pull request!!!

Nice, probably won't make changes in this PR, but I'll start using that in the future!

File: src/main/java/org/breedinginsight/services/brapi/BrAPIClientProvider.java [hunk 8]
```diff
@@ -45,4 +48,18 @@ public BrAPIClient getClient(BrAPIClientType clientType){
...
+    public List<BrAPIClient> getAllUniqueClients(){
+
+        List<BrAPIClient> clients = new ArrayList<>();
+        clients.add(coreClient);
+        if (!coreClient.brapiURI().equals(phenoClient.brapiURI())){
+            clients.add(phenoClient);
+        }
+        if (!coreClient.brapiURI().equals(genoClient.brapiURI()) && !phenoClient.brapiURI().equals(genoClient.brapiURI())) {
+            clients.add(genoClient);
+        }
```
Comment (timparsons) [inline src/main/java/org/breedinginsight/services/brapi/BrAPIClientProvider.java:60]: This could be simplified by having the return type be a `Set` (and use the `TreeSet` implementation of `Set`):

```java
Set<BrAPIClient> clients = new TreeSet<>(Comparator.comparing(BrAPIClient::brapiURI));
```

Then just add the `core`, `pheno`, and `geno` clients

File: src/main/java/org/breedinginsight/services/exceptions/ValidatorException.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+package org.breedinginsight.services.exceptions;
+
+import lombok.Getter;
+import org.breedinginsight.api.model.v1.response.ValidationErrors;
+
+@Getter
+public class ValidatorException extends Exception {
+
+    ValidationErrors errors;
```
Comment (timparsons) [inline src/main/java/org/breedinginsight/services/exceptions/ValidatorException.java:26]: This field should be `private`

File: src/main/java/org/breedinginsight/services/validators/TraitValidator.java [hunk 10]
```diff
@@ -16,70 +16,269 @@
...
 
-    public static void checkRequiredTraitFields(Trait trait) throws UnprocessableEntityException {
+    private TraitDAO traitDAO;
 
-        Method method = trait.getMethod();
-        Scale scale = trait.getScale();
+    @Inject
+    public TraitValidator(TraitDAO traitDAO){
+        this.traitDAO = traitDAO;
+    }
```
Comment (timparsons) [inline src/main/java/org/breedinginsight/services/validators/TraitValidator.java:41]: Looking at this code, I would have expected that the static methods that use the `traitDAO` wouldn't work unless the TraitValidator is injected into the class that's using it.

File: src/main/resources/brapi/properties/application.properties [hunk 11]
```diff
@@ -0,0 +1,14 @@
+application.properties.templateserver.port = 8080
```
Comment (timparsons) [inline src/main/resources/brapi/properties/application.properties:1]: Missing file copyright header

File: src/test/java/org/breedinginsight/BrAPITest.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+public class BrAPITest extends DatabaseTest {
+
+    @Getter
+    private static GenericContainer brapiContainer;
+    @Getter
+    private DSLContext brapiDsl;
+
+    @SneakyThrows
+    public BrAPITest() {
+        brapiContainer = new GenericContainer<>("breedinginsight/brapi-java-server")
```
Comment (timparsons) [inline src/test/java/org/breedinginsight/BrAPITest.java:45]: This is just awesome that there's the ability to do this now.  Love it!

File: src/test/java/org/breedinginsight/api/auth/rules/ValidUserSecurityRuleIntegrationTest.java [hunk 13]
```diff
@@ -39,9 +40,9 @@
 @MicronautTest
 @TestInstance(TestInstance.Lifecycle.PER_CLASS)
 @TestMethodOrder(MethodOrderer.OrderAnnotation.class)
-public class ValidUserSecurityRuleIntegrationTest {
+public class ValidUserSecurityRuleIntegrationTest extends DatabaseTest {
 
-    private String testUserUUID;
+    String testUserUUID;
```
Comment (timparsons) [inline src/test/java/org/breedinginsight/api/auth/rules/ValidUserSecurityRuleIntegrationTest.java:45]: Why did the `private` keyword get removed?

File: src/test/resources/brapi/properties/application.properties [hunk 14]
```diff
@@ -0,0 +1,14 @@
+application.properties.templateserver.port = 8080
```
Comment (timparsons) [inline src/test/resources/brapi/properties/application.properties:1]: Missing file copyright header

File: src/main/java/org/breedinginsight/api/model/v1/response/ValidationErrors.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+import lombok.experimental.Accessors;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Getter
+@Setter
+@Accessors(chain=true)
+@ToString
+public class ValidationErrors {
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/main/java/org/breedinginsight/api/model/v1/response/ValidationErrors.java:30]: This contains many RowValidationErrors

File: src/main/java/org/breedinginsight/daos/TraitDAO.java [hunk 16]
```diff
@@ -211,4 +356,56 @@ private Trait parseTraitRecord(Record record, BiUserTable createdByUser, BiUserT
...
+        if (valueRows.length > 0){
+            Table newTraits = dsl.select()
+                    .from(values(valueRows).as("newTraits", "new_trait_name", "new_method_name", "new_scale_name")).asTable("newTraits");
+
+            Result<Record> records = dsl.select()
+                    .from(newTraits)
+                    .join(TRAIT).on(TRAIT.TRAIT_NAME.like(newTraits.field("new_trait_name")))
+                    .join(METHOD).on(TRAIT.METHOD_ID.eq(METHOD.ID)).and(METHOD.METHOD_NAME.like(newTraits.field("new_method_name")))
+                    .join(SCALE).on(TRAIT.SCALE_ID.eq(SCALE.ID)).and(SCALE.SCALE_NAME.like(newTraits.field("new_scale_name")))
+                    .fetch();
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/main/java/org/breedinginsight/daos/TraitDAO.java:380]: Yep, that is correct. I am using the `like` operator so case doesn't matter. My thoughts are that if we have two trait names, 'Leaf Length' and 'leaf length' those are still duplicates.

File: src/main/java/org/breedinginsight/services/ProgramObservationLevelService.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+import javax.inject.Inject;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+public class ProgramObservationLevelService {
+
+    @Inject
+    private ProgramObservationLevelDAO programObservationLevelDAO;
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/ProgramObservationLevelService.java:32]: I can switch this

File: src/main/java/org/breedinginsight/services/ProgramService.java [hunk 18]
```diff
@@ -41,7 +44,13 @@
     @Inject
     private ProgramDAO dao;
     @Inject
+    private ProgramOntologyDAO programOntologyDAO;
+    @Inject
+    private ProgramObservationLevelDAO programObservationLevelDAO;
+    @Inject
     private SpeciesService speciesService;
+    @Inject
+    private DSLContext dsl;
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/ProgramService.java:53]: Yeah I would prefer to do that refactoring in another card, but I agree we need to switch everything over at some point. This likely got changed when I was changing my new work to have all private vars.

File: src/main/java/org/breedinginsight/services/TraitService.java [hunk 19]
```diff
@@ -59,4 +83,107 @@
...
+                }
+            }
+        }
+
+        // Validate the traits
+        ValidationErrors requiredFieldErrors = TraitValidator.checkRequiredTraitFields(traits);
+        ValidationErrors dataConsistencyErrors = TraitValidator.checkTraitDataConsistency(traits);
+        ValidationErrors duplicateTraits = traitValidator.checkDuplicateTraitsExisting(traits);
+        ValidationErrors duplicateTraitsInFile = TraitValidator.checkDuplicateTraitsInFile(traits);
+        validationErrors.mergeAll(requiredFieldErrors, dataConsistencyErrors, duplicateTraits, duplicateTraitsInFile);
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/TraitService.java:136]: Yep, that's a good idea. I'll make that change.

File: src/main/java/org/breedinginsight/services/brapi/BrAPIClientProvider.java (hunk 8 above)
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/brapi/BrAPIClientProvider.java:60]: Ah cool, I was wondering if you could customize the set comparator. I'll change this.

File: src/main/java/org/breedinginsight/services/exceptions/ValidatorException.java (hunk 9 above)
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/exceptions/ValidatorException.java:26]: I knew I'd miss at least one...

File: src/test/java/org/breedinginsight/api/auth/rules/ValidUserSecurityRuleIntegrationTest.java (hunk 13 above)
Comment (ctucker3) [reply src/test/java/org/breedinginsight/api/auth/rules/ValidUserSecurityRuleIntegrationTest.java:45]: Hrmm not sure, I'll change it back.

File: src/main/java/org/breedinginsight/services/validators/TraitValidator.java (hunk 10 above)
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/validators/TraitValidator.java:41]: Not sure I'm following on this one. None of the static methods use traitDao. Maybe we can chat about this so you can show me what you are referencing?

File: docs/openapi.yaml [hunk 20]
```diff
@@ -3685,6 +3631,33 @@ components:
...
+            type: object
+            properties:
+              rowIndex:
+                type: integer
+              errors:
+                type: array
+                items:
+                  type: object
+                  properties:
+                    column:
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply docs/openapi.yaml:3649]: Yeah for the POST request that makes sense to me. Thoughts on making it "field" for the trait file upload as well? It will be easier that way

File: src/main/java/org/breedinginsight/services/TraitService.java (hunk 7 above)
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/TraitService.java:148]: Changed and pushed

File: README.md [hunk 21]
```diff
@@ -17,10 +17,9 @@ to be run in: e.g develop, test, staging, and production.  Each service contains
 under the environment key public values for environment variables used as params
 for the API configuration.
 
-Private values used in each environment are stored in Lastpass and are never
-placed in docker-compose.yml and never committed to the repo.  At the root level
-of the repo locally create a file called .env and save the Lastpass contents for
-"bi-api secrets" in this file.
+Private values used in each environment are never placed in docker-compose.yml 
+and never committed to the repo. Variables are passed into the docker-compose.yml
+from the .env file. The .env file should be modified to contain the secrets.
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply README.md:22]: One thing I found with the different env files is that it has to specified for each service, and can't be a one line change to the docker-compose file for all the services. You also can't specify the new env file in the command line. I wish docker-compose made it easier to do this!

File: .env (hunk 1 above)
Comment (ctucker3) [reply .env:31]: I tried to do `http://localhost:${BRAPI_SERVER_PORT}` for this, but I got micronaut errors because of it. I think we may have to leave it has localhost:8083 for now and revisit it later if it causes too many issues.

File: src/main/java/org/breedinginsight/services/ProgramObservationLevelService.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/ProgramObservationLevelService.java]: Changed and pushed. I changed all of the main classes for this workflow.

File: src/main/java/org/breedinginsight/services/ProgramOntologyService.java [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+
+import javax.inject.Inject;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+public class ProgramOntologyService {
+
+    @Inject
+    private ProgramOntologyDAO programOntologyDAO;
```
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/ProgramOntologyService.java]: Changed and pushed

File: src/main/java/org/breedinginsight/services/TraitService.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/TraitService.java]: Changed and pushed. I wasn't able to do this exactly the way described in your comment because of the program observation level check above this block of code. I did combine all of these validations into one method call though.

File: src/main/java/org/breedinginsight/services/brapi/BrAPIClientProvider.java (hunk 8 above)
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/brapi/BrAPIClientProvider.java]: Changed and pushed

File: src/main/java/org/breedinginsight/services/exceptions/ValidatorException.java (hunk 9 above)
Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/exceptions/ValidatorException.java]: Changed and pushed

File: src/main/resources/brapi/properties/application.properties (hunk 11 above)
Comment (ctucker3) [reply src/main/resources/brapi/properties/application.properties:18]: Changed and pushed

File: src/test/java/org/breedinginsight/api/auth/rules/ValidUserSecurityRuleIntegrationTest.java (hunk 13 above)
Comment (ctucker3) [reply src/test/java/org/breedinginsight/api/auth/rules/ValidUserSecurityRuleIntegrationTest.java]: Changed and pushed

File: src/test/resources/brapi/properties/application.properties (hunk 14 above)
Comment (ctucker3) [reply src/test/resources/brapi/properties/application.properties:18]: Changed and pushed

File: docs/openapi.yaml (hunk 2 above)
Comment (ctucker3) [reply docs/openapi.yaml]: Changed and pushed

File: docs/openapi.yaml (hunk 3 above)
Comment (ctucker3) [reply docs/openapi.yaml]: Changed and pushed

File: README.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (ctucker3) [reply README.md:22]: Nevermind, Tim sent me the command he had and it worked great! I was using the wrong syntax. The instructions for this will be added to the PR for ONT-68.

File: docs/openapi.yaml (hunk 20 above)
[the remark this answers is not in the document]
Comment (ctucker3) [reply docs/openapi.yaml]: Updated both trait-upload endpoints and the POST traits endpoint to have `field` instead of `column` in the error body.

File: src/main/java/org/breedinginsight/services/TraitService.java [hunk 23]
```diff
@@ -59,4 +93,158 @@
...
+        // Get our program observation levels
+        List<ProgramObservationLevel> programLevels = programObservationLevelService.getByProgramId(programId);
+        for (int i = 0; i < traits.size(); i++) {
+            Trait trait = traits.get(i);
+            if (trait.getProgramObservationLevel() != null){
+                List<ProgramObservationLevel> matchingLevels = programLevels.stream()
+                        .filter(programObservationLevel -> programObservationLevel.getName().equals(trait.getProgramObservationLevel().getName()))
+                        .collect(Collectors.toList());
+                if (matchingLevels.size() == 0) {
+                    ValidationError validationError = traitValidatorError.getTraitLevelDoesNotExist();
```
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/services/TraitService.java:236]: Talked about adding list of available levels to error message so user can know what is a valid level.

Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/TraitService.java]: Changed and pushed
